### PR TITLE
Add suport for Consul Connect

### PR DESCRIPTION
**Pull request by Gufran**
### What does this PR do?

The change set introduces support for [Consul Connect][] enabled services.

### Motivation

There is no edge proxy available that can route traffic to a connect enabled service. Consul Connect, despite being a powerful and easy to use service mesh, is useless to a lot of people who are mainly looking to route traffic from internet to private services. A service running inside Connect service mesh can only receive traffic via its sidecar, and sidecar will only communicate with a network peer using mutual TLS. The solution is easy, but haven't been implemented in any form.

Traefik already supports Consul Catalog, it is only a matter of utilizing the certificates for upstream connection wherever applicable and it becomes the perfect edge proxy for connect mesh.

### More

- [ ] Added/updated tests
- [ ] Added/updated documentation

### Additional Notes

This PR is in progress, I need some help to figure out how to set the TLS configuration on a connection without specifying it in service tags.

Related: https://github.com/containous/traefik/issues/3544

[Consul Connect]: https://www.consul.io/docs/connect/index.html

### CODE REVIEW

File: pkg/provider/consulcatalog/config.go [hunk 1]
```diff
@@ -33,6 +34,10 @@ func (p *Provider) buildConfiguration(ctx context.Context, items []itemData) *dy
 			continue
 		}
 
+		if item.ExtraConf.ConnectEnabled {
+			confFromLabel.TLS = tlsFromLeafCert(tlsInfo)
```
Comment (Gufran) [inline pkg/provider/consulcatalog/config.go]: apparently this is not enough to propagate the certificates all the way up to the connection.
any help here would be greatly appreciated.

**Comment by Gufran** [comment]
Testing the behaviour:
in [test.zip](https://github.com/containous/traefik/files/4250735/test.zip) there are 3 files `test.sh`, `config.d/service-with-connect.json`, and `config.d/service-without-connect.json`.

Inflate the archive, download `http-echo` service (`go get github.com/hashicorp/http-echo`) and run `test.sh` to start consul, two services and a connect sidecar proxy.

Traefik will be able to route to service `without-connect` but not to service `with-connect`. The connection to sidecar fails with a `502`.

**Comment by Gufran** [comment]
@ldez can you please help me here? I am trying to configure TLS on the router dynamically.

**Comment by mmatur** [comment]
Hi @Gufran,

Thanks a lot for this contribution.

Unfortunately I think you will be blocked until an implementation for #4835 has been done. 

Today it is not possible to create a serverTransport per service and it is something that is required if we want to be able to create mTLS connection with others consul service.

**Comment by Gufran** [comment]
@mmatur thanks for the update. that changes things a bit.
I'll close this PR and implement the support for Consul Connect in a way that does not require configuring TLS from within the provider.

**Comment by the-maldridge** [comment]
Is there an updated PR somewhere that I can follow for this?

**Comment by Gufran** [comment]
No, I had to give up on this feature because the core does not support creating mTLS connections with upstreams.

I ended up creating https://github.com/Gufran/flightpath though, check it out if you are looking for a router for consul connect.

**Comment by blake** [comment]
We just announced the beta release of Consul 1.8 which includes native support for Ingress gateways using Envoy.

See the announcement blog for more details (https://www.hashicorp.com/blog/announcing-hashicorp-consul-1-8/), or head over to https://www.consul.io/docs/connect/ingress_gateway for info configuring a gateway.

**Comment by apollo13** [comment]
Hi @Gufran, #4835 is merged now - do you think this would make integration in traefik easier? Do you have any interest in updating your PR?

**Comment by jharley** [comment]
> Hi @Gufran, #4835 is merged now - do you think this would make integration in traefik easier? Do you have any interest in updating your PR?

@Gufran @apollo13 I am interested in helping however I can to get this support into Traefik now that per-Service Transports are supported 🚀

**Comment by Gufran** [comment]
Just glanced over #4835 and it appears to be a straightforward configuration. I might be able to pick up this PR again over this weekend. Lets see how that pans out.

**Comment by apollo13** [comment]
@Gufran Did you hit any roadblocks? Is there still something missing in traefik to get this implemented?

**Comment by Gufran** [comment]
@apollo13 My apologies! I did hit a couple of snags with package names and generated code, apparently the repository has been renamed from `continuous/traefik` to `traefik/traefik` since this PR. That is sorted, and I was able to update the code to utilize the transport but I didn't get enough time to write tests or test it out otherwise. It looks like there is still some problem with mTLS verification and Traefik is still getting error from connect sidecar.

I can't reinstate this PR for some reason so I've created a new one https://github.com/traefik/traefik/pull/7407

I'll try to set out some time over the weekend and debug it, and also write the test if I am able to fix the problem.

**Comment by apollo13** [comment]
Hi @Gufran, no need to apologize, I highly appreciate you working on it! Sadly my go skills are close to non-existent, so all I can do is pester other people :) But if you need people to test I'll happily do so once you figured out the kinks -- just let me know if I can do anything to help.

**Pull request closed without merging**